pkg/cloud-init/cloud-init.go

```diff
@@ -82,20 +83,73 @@ type PublicSSHKey struct {
 }
 
 type NoCloudMetadata struct {
-	InstanceType  string            `json:"instance-type,omitempty"`
-	InstanceID    string            `json:"instance-id"`
-	LocalHostname string            `json:"local-hostname,omitempty"`
-	PublicSSHKeys map[string]string `json:"public-keys,omitempty"`
+	InstanceType   string            `json:"instance-type,omitempty"`
+	InstanceID     string            `json:"instance-id"`
+	LocalHostname  string            `json:"local-hostname,omitempty"`
+	PublicSSHKeys  map[string]string `json:"public-keys,omitempty"`
+	CustomMetadata map[string]string `json:"-"`
+}
+
+var reservedNoCloudFields = map[string]struct{}{
+	"instance-type":  {},
+	"instance-id":    {},
+	"local-hostname": {},
+	"public-keys":    {},
+}
+
+func (ncm *NoCloudMetadata) MarshalJSON() ([]byte, error) {
+	type noCloudMetadata NoCloudMetadata
+	alias := (*noCloudMetadata)(ncm)
+	return marshalMetadataWithCustomFields(alias, ncm.CustomMetadata, reservedNoCloudFields)
 }
 
 type ConfigDriveMetadata struct {
-	InstanceType  string            `json:"instance_type,omitempty"`
-	InstanceID    string            `json:"instance_id"`
-	LocalHostname string            `json:"local_hostname,omitempty"`
-	Hostname      string            `json:"hostname,omitempty"`
-	UUID          string            `json:"uuid,omitempty"`
-	Devices       *[]DeviceData     `json:"devices,omitempty"`
-	PublicSSHKeys map[string]string `json:"public_keys,omitempty"`
+	InstanceType   string            `json:"instance_type,omitempty"`
+	InstanceID     string            `json:"instance_id"`
+	LocalHostname  string            `json:"local_hostname,omitempty"`
+	Hostname       string            `json:"hostname,omitempty"`
+	UUID           string            `json:"uuid,omitempty"`
+	Devices        *[]DeviceData     `json:"devices,omitempty"`
+	PublicSSHKeys  map[string]string `json:"public_keys,omitempty"`
+	CustomMetadata map[string]string `json:"-"`
+}
+
+var reservedConfigDriveFields = map[string]struct{}{
+	"instance_type":  {},
+	"instance_id":    {},
+	"local_hostname": {},
+	"hostname":       {},
+	"uuid":           {},
+	"devices":        {},
+	"public_keys":    {},
+}
+
+func (cmd *ConfigDriveMetadata) MarshalJSON() ([]byte, error) {
+	type configDriveMetadata ConfigDriveMetadata
+	alias := (*configDriveMetadata)(cmd)
+	return marshalMetadataWithCustomFields(alias, cmd.CustomMetadata, reservedConfigDriveFields)
+}
+
+func marshalMetadataWithCustomFields(metadata interface{}, customMetadata map[string]string, reservedFields map[string]struct{}) ([]byte, error) {
+	baseMap := make(map[string]interface{})
+	baseBytes, err := json.Marshal(metadata)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(baseBytes, &baseMap); err != nil {
+		return nil, err
+	}
+
+	for key, value := range customMetadata {
+		if _, isReserved := reservedFields[key]; isReserved {
+			log.Log.V(2).Infof("Custom metadata key '%s' conflicts with reserved field and will be skipped", key)
```

Comment:
Thinking about this again, it might be better to return error instead of silently swallowing duplicates.